common/diagnostics/Metrics.java

```diff
@@ -6,203 +6,752 @@
 
 package com.vaticle.typedb.core.common.diagnostics;
 
+import com.eclipsesource.json.JsonArray;
 import com.eclipsesource.json.JsonObject;
-
-import java.util.TimeZone;
+import com.vaticle.typedb.common.collection.ConcurrentSet;
+import com.vaticle.typedb.core.common.parameters.Arguments;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.lang.management.ManagementFactory;
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicLong;
 
+import static com.vaticle.typedb.core.common.diagnostics.StatisticReporter.REPORT_INTERVAL_MINUTES;
+import static com.vaticle.typedb.core.common.diagnostics.Version.JSON_API_VERSION;
+
 public class Metrics {
-    private final SystemProperties system;
-    private final NetworkRequests requests;
-    private final CurrentCounts usage;
-    private final UserErrorStatistics userErrors;
+    private final ConcurrentSet<String> primaryDatabaseHashes = new ConcurrentSet<>();
+    private final BaseProperties base;
+    private final ServerProperties serverProperties;
+    private final ConcurrentMap<String, DatabaseLoadDiagnostics> databaseLoad = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, NetworkRequests> requests = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, UserErrorStatistics> userErrors = new ConcurrentHashMap<>();
+
+    public Metrics(String deploymentID, String serverID, String name, String version, boolean reportingEnabled, Path dataDirectory) {
+        this.base = new BaseProperties(JSON_API_VERSION, deploymentID, serverID, name, reportingEnabled);
+        this.serverProperties = new ServerProperties(version, dataDirectory);
+    }
 
-    public Metrics(String serverID, String name, String version) {
-        this.system = new SystemProperties(serverID, name, version);
-        this.requests = new NetworkRequests();
-        this.usage = new CurrentCounts();
-        this.userErrors = new UserErrorStatistics();
+    public void takeSnapshot() {
+        this.base.updateSinceTimestamp();
+        this.databaseLoad.values().forEach(DatabaseLoadDiagnostics::takeSnapshot);
+        this.requests.values().forEach(NetworkRequests::takeSnapshot);
+        this.userErrors.values().forEach(UserErrorStatistics::takeSnapshot);
     }
 
-    public void requestAttempt(Metrics.NetworkRequests.Kind kind) {
-        requests.attempt(kind);
+    private String hashAndAddDatabaseIfAbsent(@Nullable String databaseName) {
+        String databaseHash = databaseName != null ? String.valueOf(databaseName.hashCode()) : "";
```

Comment:
I use java's hash code to hash database names instead of a smarter algorithm here as I believe it is enough for our mission: to hide the real name and not be able to reverse this hashing, to have consistent hashes for the same values and not to have collisions within **one deployment**. I don't believe that a deployment can have hundreds of thousands databases and more, so I'm not scared of collisions here.